core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -302,16 +260,23 @@ private[spark] class TaskSetManager(
   private def dequeueTaskFromList(
       execId: String,
       host: String,
-      list: ArrayBuffer[Int]): Option[Int] = {
+      list: ArrayBuffer[Int],
+      speculative: Boolean = false): Option[Int] = {
     var indexOffset = list.size
     while (indexOffset > 0) {
       indexOffset -= 1
       val index = list(indexOffset)
-      if (!isTaskBlacklistedOnExecOrNode(index, execId, host)) {
+      if (!isTaskBlacklistedOnExecOrNode(index, execId, host) &&
+          !(speculative && hasAttemptOnHost(index, host))) {
         // This should almost always be list.trimEnd(1) to remove tail
         list.remove(indexOffset)
-        if (copiesRunning(index) == 0 && !successful(index)) {
-          return Some(index)
+        if (!successful(index)) {
+          if (copiesRunning(index) == 0) {
+            return Some(index)
+          } else if (speculative && copiesRunning(index) == 1) {
+            speculatableTasks -= index
```

Comment:
when `copiesRunning(index) == 0`, it is still possible to return a speculatable tasks, thinking about the case that a long running task failed after it got a pending  speculatable task. I mean,  for `if (copiesRunning(index) == 0)` above, it actually covers two cases:  1) not speculative 2) speculative. So, I think it's better to put this after `dequeueTaskHelper(execId, host, maxLocality, true)`.

Reply:
Makes sense, have updated the code.